mangaki/mangaki/utils/dpp.py

```diff
@@ -0,0 +1,129 @@
+from sklearn.utils.extmath import randomized_svd
+from scipy.spatial.distance import pdist, squareform
+from numpy.random import choice
+import numpy as np
+
+MAX_ITER_SAMPLE_DPP = 10
+
+
+def diameter(r, points):
+    nb_points = points.shape[0]
+    return ((2 / (nb_points * (nb_points - 1)) *
+            ((pdist(points)**r).sum()))**(1 / r))
+
+
+def diameter_0(points):
+    r = 1
+    first = diameter(r, points)
+    second = diameter(r / 2, points)
+    while (first - second) > (0.01 * second):
+        first = diameter(r, points)
+        r = r / 2
+        second = diameter(r, points)
+    return second
+
+
+class SimilarityMatrix:
+
+    def __init__(self, matrix, nb_components_svd=10,
+                 fname=None, algo='svd', metric='cosine'):
+        self.nb_components_svd = nb_components_svd
+        self.algo = algo
+        self.matrix = matrix
+        self.similarity_matrix = self.make_similarity_matrix(metric)
+
+    def make_svd_matrix(self):
+        self.U, self.sigma, self.VT = randomized_svd(
+            self.matrix, self.nb_components_svd)
+
+    def make_similarity_matrix(self, metric):
+        if self.algo == 'svd':
+            self.make_svd_matrix()
+            return 1 - squareform(pdist(self.VT.T, metric=metric))
+        return 1 - squareform(pdist(self.matrix.T, metric=metric))
+
+
+class MangakiUniform:
+
+    def __init__(self, items):
+        self.items = items
+
+    def sample_k(self, nb_points):
+        return choice(self.items, nb_points).tolist()
+
+
+class MangakiDPP:
+
+    def __init__(self, items, similarity_matrix):
+        self.items = items
+        self.similarity_matrix = similarity_matrix
+
+    def sample_k(self, *args, **kwargs):
+        for i in range(MAX_ITER_SAMPLE_DPP):
+            try:
+                return self._sample_k(*args, **kwargs)
+            except np.linalg.linalg.LinAlgError as e:
+                print('LinAlgError in MangakiDPP')
+        raise ValueError('Too much LinAlgError')
```

Comment:
iirc, c'est un problème avec la matrice utilisée qui n'est pas forcément inversible donc c'est peut-être plus logique de laisser tel quel. Après JJ avait proposé une autre implémentation trouvée sur GitHub et normalement j'aurais le temps de la voir d'ici le WE prochain